lib/internal/streams/pipeline.js

```diff
@@ -73,25 +75,129 @@ function pipeline(...streams) {
+  const destroys = [];
+
+  function finish(err, final) {
+    if (!error) error = err;
+    if (err) {
+      for (const destroy of destroys) {
+        destroy(err);
+      }
+    }
+    if (!final) return;
+    for (const destroy of destroys) {
+      destroy();
+    }
+    callback(error);
+  }
+
+  async function pump(iterableOrPromise, writable) {
+    try {
+      const iterable = await iterableOrPromise;
+
+      if (iterable[SymbolAsyncIterator]) {
+        for await (const chunk of iterable) {
+          if (writable.destroyed) return;
+          if (!writable.write(chunk)) await events_once(writable, 'drain');
+          if (writable.destroyed) return;
+        }
+      } else if (iterable[SymbolIterator]) {
+        for (const chunk of iterable) {
+          if (!writable.write(chunk)) await events_once(writable, 'drain');
+          if (writable.destroyed) return;
         }
+      } else {
+        throw new ERR_INVALID_ARG_TYPE(
+          'iterable', ['Iterable', 'AsyncIterable'], iterable);
       }
-      if (reading) return;
-      for (const destroy of destroys) {
-        destroy();
+      writable.end();
+    } catch (err) {
+      finish(err, false);
+    }
+  }
+
+  function wrap(stream, reading, writing) {
+    destroys.push(destroyer(stream, reading, writing, function(err) {
+      finish(err, !reading);
+    }));
+  }
+
+  function makeAsyncIterable(iterableOrPromise) {
+    return async function*() {
+      const iterable = await iterableOrPromise;
+      if (iterable[SymbolAsyncIterator]) {
+        for await (const chunk of iterable) {
+          yield chunk;
+        }
+      } else if (iterable[SymbolIterator]) {
+        for (const chunk of iterable) {
+          yield chunk;
+        }
+      } else {
+        throw new ERR_INVALID_ARG_TYPE(
+          'iterable', ['Iterable', 'AsyncIterable'], iterable);
       }
-      callback(error);
-    });
-  });
+    }();
+  }
+
+  let ret;
+  for (let i = 0; i < streams.length; i++) {
+    const stream = streams[i];
+    if (i === 0 && !stream.pipe &&
+        (stream[SymbolIterator] || stream[SymbolAsyncIterator])) {
+      ret = stream;
+    } else if (typeof stream === 'function') {
+      if (ret && (ret.then || ret[SymbolIterator])) {
+        ret = makeAsyncIterable(ret);
+      }
+      ret = stream(ret);
+
+      // destination
+      if (i === streams.length - 1) {
+        if (ret.pipe || ret.write) {
+          // Do nothing...
+        } else if (ret[SymbolAsyncIterator]) {
+          ret = async function*() {
+            try {
+              for (const chunk of ret) {
+                yield chunk;
+              }
+            } catch (err) {
+              finish(err, true);
+              throw err;
+            }
+          };
+        } else if (ret.then) {
+          const fn = (err) => {
+            finish(err, true);
+          };
+
+          ret
+            .then(fn)
+            .catch(fn);
+
+          // TODO(ronag): return promise?
```

Comment:
return promise or not? Problem with returning a promise is that you must attach a rejection handler... which kind of conflicts with the `callback` semantics.